src/computational_graph/transform.jl

```diff
@@ -87,21 +81,121 @@ end
 
-This function first finds all labels involved in g and its subgraphs (for example, 1, 4, 5, 7, ...), 
-then relabel them in the order 1, 2, 3, 4, ....
-Return a copy of stantardized g.
+    This function first finds all labels involved in g and its subgraphs (for example, 1, 4, 5, 7, ...), 
+    then relabel them in the order 1, 2, 3, 4, ....
+    Returns a standardized copy of g.
 """
 standardize_labels(g::Graph) = standardize_labels!(deepcopy(g))
 
-############LEGACY BELOW################
+"""
+    function replace_subgraph!(g::Graph, w::Graph, m::graph)
 
-# function readDiag(io::IO)
+    In place function that replaces the children graph w in graph g with a new graph m.
+    Graphs w and m should have the same internal and external vertices, and topology
+"""
+function replace_subgraph!(g::Graph, w::Graph, m::Graph)
+    @assert !isleaf(g) "Target parent graph can not be a leaf"
+    @assert w.vertices == m.vertices "Old and new subgraph should have the same vertices"
+    @assert w.external == m.external "Old and new subgraph should have the same external vertices"
+    print("isleaf $(isleaf(g))\n")
+    for node in PreOrderDFS(g)
+        for (i, child) in enumerate(children(node))
+            if isequiv(child, w, :id)
+                node.subgraphs[i] = m
+                return
+            end
+        end
+    end
+end
+
+"""
+    function replace_subgraph(g::Graph, w::Graph, m::graph)
 
-#     return Diagram
-# end
+    Generate a copy of graph g, with the children graph w replaced by a new graph m.
+    Graph w and m should have the same internal and external vertices, and topology
+"""
+function replace_subgraph(g::Graph, w::Graph, m::Graph)
+    @assert w.vertices == m.vertices "Old and new subgraph should have the same vertices"
+    @assert w.external == m.external "Old and new subgraph should have the same external vertices"
+    g0 = deepcopy(g)
+    for node in PreOrderDFS(g0)
+        for (i, child) in enumerate(children(node))
+            if isequiv(child, w, :id)
+                node.subgraphs[i] = m
+                break
+            end
+        end
+    end
+    return g0
+end
 
-# function printDiag(diag::Union{Tuple,AbstractVector}, kwargs...)
+"""
+    function prune_trivial_unary(g::Graph)
 
-#     f = open("./DiagFiles/Diag.txt", "w")
-#     writedlm(f, [])
-# end
+    Returns a simplified copy of g if it represents a trivial unary operation.
+    Otherwise, returns the original graph.
+"""
+function prune_trivial_unary(g::Graph)
+    # No-op; g is not a branch (depth-1, one-child tree)
+    if isbranch(g) == false
+        return g
+    end
+    # Prune trivial unary operation
+    if unary_istrivial(g.operator) && isfactorless(g)
+        return eldest(g)
+    else
+        return g
+    end
+end
+
+"""
+    function inplace_prod(g::Graph)
+
+    Converts a unary Prod link to in-place form by propagating subgraph_factors up a level.
+"""
+function inplace_prod(g::Graph)
```

Comment:
why inplace_prod only search for one level?

Reply:
`Base.*(g::Graph{F,W}, c::C)` already builds in a one-level inplace_prod (we should update that code to actually call the function). Hence, I think we can only form a Prod chain of length at most 2 during future transforms like expand, replace_subgraph, etc.

Reply:
Fixed